include/aspect/stokes_matrix_free.h

```diff
@@ -0,0 +1,652 @@
+/*
+  Copyright (C) 2018 - 2019 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef _aspect_stokes_matrix_free_h
+#define _aspect_stokes_matrix_free_h
+
+#include <aspect/global.h>
+
+#include <aspect/simulator.h>
+
+#include <deal.II/matrix_free/matrix_free.h>
+#include <deal.II/matrix_free/operators.h>
+#include <deal.II/matrix_free/fe_evaluation.h>
+
+#include <deal.II/multigrid/mg_constrained_dofs.h>
+#include <deal.II/multigrid/multigrid.h>
+#include <deal.II/multigrid/mg_transfer_matrix_free.h>
+#include <deal.II/multigrid/mg_tools.h>
+#include <deal.II/multigrid/mg_coarse.h>
+#include <deal.II/multigrid/mg_smoother.h>
+#include <deal.II/multigrid/mg_matrix.h>
+
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/block_vector.h>
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/la_parallel_block_vector.h>
+
+namespace aspect
+{
+  using namespace dealii;
+
+  namespace MatrixFreeStokesOperators
+  {
+
+    template <int dim, int degree_v, typename number>
+    class StokesOperator
+      : public MatrixFreeOperators::Base<dim, dealii::LinearAlgebra::distributed::BlockVector<number> >
+    {
+      public:
+        StokesOperator ();
+        void clear ();
+        void fill_viscosities_and_pressure_scaling(const dealii::LinearAlgebra::distributed::Vector<number> &visc_vals,
+                                                   const double scaling,
```

Comment:
can some of these made private?

Reply:
Both `compute_diagonal()` and `fill_viscosities_and_pressure_scaling()` must (I believe) be public since we call them in `stokes_matrix_free.cc`. We never actually use `clear()' so maybe that can be moved. Or it may not be needed at all?